src/main/java/com/marklogic/kafka/connect/source/SerializedConstraintColumnHandler.java

```diff
@@ -0,0 +1,92 @@
+import com.marklogic.client.io.StringHandle;
+import com.marklogic.client.row.RawPlanDefinition;
+import com.marklogic.client.row.RowManager;
+
+import java.util.Map;
+
+public class SerializedConstraintColumnHandler extends LoggingObject implements ConstraintColumnHandler {
+    private final ObjectMapper mapper = new ObjectMapper();
+    private final DatabaseClient databaseClient;
+
+    SerializedConstraintColumnHandler(DatabaseClient databaseClient) {
+        this.databaseClient = databaseClient;
+    }
+
+    @Override
+    public String injectConstraintIntoQuery(Map<String, Object> parsedConfig, String constraintColumn, String previousMaxConstraintColumnValue) {
+        String originalQuery = (String) parsedConfig.get(MarkLogicSourceConfig.SERIALIZED_QUERY);
+        String constrainedQuery = originalQuery;
+        if (previousMaxConstraintColumnValue != null) {
+            String constraintNodeString = "{\"ns\":\"op\", \"fn\":\"where\", \"args\":[{\"ns\":\"op\", \"fn\":\"gt\", \"args\":[{\"ns\":\"op\", \"fn\":\"col\", \"args\":[\"" + constraintColumn + "\"]}, \"" + previousMaxConstraintColumnValue + "\"]}]}";
+            try {
+                ObjectNode constraintNode = (ObjectNode) mapper.readTree(constraintNodeString);
+
+                ObjectNode queryRoot = (ObjectNode) mapper.readTree(originalQuery);
+                ArrayNode rootArgsArray = (ArrayNode) queryRoot.get("$optic").get("args");
+
+                rootArgsArray.insert(1, constraintNode);
+                constrainedQuery = mapper.writeValueAsString(queryRoot);
+            } catch (JsonProcessingException e) {
+                throw new RuntimeException(e);
+            }
+        }
+        logger.debug("Serialized constrainedQuery (unless initial run): " + constrainedQuery);
+        return constrainedQuery;
+    }
+
+    @Override
+    public String getMaxConstraintValue(String currentQuery, long queryStartTimeInMillis, String constraintColumn) {
+        String currentMaxValueQuery = buildMaxValueSerializedQuery(currentQuery, constraintColumn);
+        logger.debug("currentMaxValueQuery: " + currentMaxValueQuery);
+        RowManager rowMgr = databaseClient.newRowManager();
+        RawPlanDefinition maxConstraintValueQuery = rowMgr.newRawPlanDefinition(new StringHandle(currentMaxValueQuery));
+        JacksonHandle handle = new JacksonHandle().withFormat(Format.JSON).withMimetype("application/json");
+        handle.setPointInTimeQueryTimestamp(queryStartTimeInMillis);
+        JacksonHandle result = rowMgr.resultDoc(maxConstraintValueQuery, handle);
+        String newMaxConstraintValue = "";
+        try {
+            newMaxConstraintValue = result.get().get("rows").get(0).get("constraint").get("value").asText();
+            newMaxConstraintValue = newMaxConstraintValue.replaceAll("[\"']", "");
+        } catch (Exception ex) {
+            logger.warn("Failed to get a valid Maximum Constraint value: " + ex.getMessage());
```

Comment:
This is an interesting error scenario. If we fail to get the max value but we don't throw an error and instead return N rows, then on the next poll(), those same N rows will be returned. I am thinking the default should be to throw an error (which will be caught by `poll()`) so that no rows are returned, but I'm not 100% sure about that. It may need to be a new config option, which would be a new story - i.e. whether the user wants an error to be thrown or not in this scenario.

Reply:
Since 1) it's conceivable that this could fail, but be recoverable (network issue) and 2) the customer could be okay with duplicate records,
Then, I lean toward defaulting to a log message (as above), but creating a config value.